# ML-04 — Search Intelligence Data Contract

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [7]:
import pandas as pd
import io

data = """client_hash_id,is_active,has_gsc_access,has_ga4_access,access_profile,client_created_date,client_updated_date,gsc_data_start,ga4_data_start
client_04660893ae39614a,true,true,true,gsc_and_ga4,2026-04-15,2026-06-27,,2026-05-22
client_05475c07ed21a83a,true,false,false,no_search_or_analytics_access,2026-04-01,2026-06-27,,
client_06d356715a8ff3b6,true,true,true,gsc_and_ga4,2026-03-23,2026-07-05,2026-04-10,2026-04-06
client_0797ff3a1fc9a6a5,true,false,false,no_search_or_analytics_access,2025-05-26,2026-06-27,2025-11-05,
client_08a6a72ff48e62c0,true,true,false,gsc_only,2025-05-26,2026-06-27,2025-09-24,
client_08d2847f24cf89c1,true,true,true,gsc_and_ga4,2025-07-11,2026-06-28,2025-07-21,2026-02-19
client_0b245132bb722950,false,true,true,gsc_and_ga4,2026-02-25,2026-06-29,2026-04-12,2026-04-24
client_0e1acc6cd57b0eba,true,true,true,gsc_and_ga4,2025-06-25,2026-06-27,2025-09-24,2026-02-19
client_0fa64a184f18a4a0,false,true,true,gsc_and_ga4,2026-01-26,2026-06-29,2026-02-19,2026-02-17
client_123b42d7ca0e1690,true,false,false,no_search_or_analytics_access,2026-06-29,2026-07-02,,
client_157ffe4d4a595515,false,true,true,gsc_and_ga4,2025-07-29,2026-07-05,2026-02-19,2026-03-09
client_15bc12546a4e861b,false,false,false,no_search_or_analytics_access,2025-05-26,2026-06-29,,
client_19b89ee4fe3db6da,false,true,true,gsc_and_ga4,2025-06-12,2026-06-29,,2026-01-09
client_1a730cb2640a1abf,true,true,true,gsc_and_ga4,2026-01-16,2026-06-27,2026-02-19,2026-03-05
client_1a8bf67cad4ee525,true,true,true,gsc_and_ga4,2026-04-21,2026-06-27,2026-05-11,2026-05-16
client_1d09b519bdde7c7a,,,,source_only_missing_client_dimension,,,2025-11-05,
client_20259bd6705d81d4,false,true,true,gsc_and_ga4,2025-10-02,2026-06-29,2026-02-19,2026-03-03
client_2094c6eb080311d5,true,true,true,gsc_and_ga4,2025-11-27,2026-06-27,2025-12-17,2025-12-18
client_230496ee2302be94,true,false,false,no_search_or_analytics_access,2026-06-29,2026-07-03,,
client_23a62021009f63c4,true,true,true,gsc_and_ga4,2025-07-21,2026-06-27,2025-09-24,2025-10-29
client_2910fd937f0b4d9a,true,true,true,gsc_and_ga4,2025-06-24,2026-06-27,2025-09-24,2026-02-19
client_2b29c545d7e25a1a,true,false,false,no_search_or_analytics_access,2026-01-16,2026-06-27,,
client_2b4306c3ed003f01,true,true,false,gsc_only,2025-09-02,2026-06-27,2026-02-19,
client_2c32078d69f2cbad,false,false,false,no_search_or_analytics_access,2025-07-28,2026-06-29,2025-11-05,
client_2c8dbc95f95ba9e6,true,false,false,no_search_or_analytics_access,2025-07-01,2026-06-27,,
client_2e65897d94f60220,,,,source_only_missing_client_dimension,,,2025-11-05,
client_3197e6291363b4db,false,true,true,gsc_and_ga4,2025-05-26,2026-06-29,2025-06-29,2025-11-09
client_3d7bd1501a0634d8,true,false,false,no_search_or_analytics_access,2025-07-09,2026-06-27,,
client_3f0ce4d44fe94f3d,true,true,true,gsc_and_ga4,2025-11-20,2026-06-27,2026-02-19,2026-03-09
client_3ffa76342f366962,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-10-11,2026-03-11
client_400c21c81c8b46ef,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-10-26,
client_46703b915c4762e0,true,false,false,no_search_or_analytics_access,2025-05-26,2026-06-27,,
client_4a18d1793d92fb84,true,true,true,gsc_and_ga4,2025-06-24,2026-06-27,2025-09-24,2026-02-19
client_4a58d657ef751d68,false,false,false,no_search_or_analytics_access,2025-05-26,2026-06-29,,
client_4bc2b415cf2ba0ea,true,false,false,no_search_or_analytics_access,2025-09-29,2026-06-27,,
client_5781daf723188fe7,true,true,false,gsc_only,2025-08-01,2026-06-27,,
client_580585c1d84137bd,true,false,false,no_search_or_analytics_access,2025-07-09,2026-06-27,,
client_59256b0571e0c970,true,true,true,gsc_and_ga4,2025-06-24,2026-06-27,2025-09-24,2026-02-19
client_599043c0ff13edea,true,true,false,gsc_only,2025-06-24,2026-06-27,2025-09-27,
client_6014ebddaa96367e,false,false,false,no_search_or_analytics_access,2026-04-16,2026-06-29,,
client_625b6439094e23e4,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-07-01,2026-02-19
client_62f4a7e64f5e0096,true,true,false,gsc_only,2025-05-26,2026-06-27,2025-06-07,
client_65de48885f4ef01b,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-06-21,2026-02-19
client_712f3d8bcd8d4cf8,true,false,false,no_search_or_analytics_access,2025-05-26,2026-06-27,,
client_73cda7b4e4f265ea,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-02-11,2026-03-24
client_764ae36a94e30a25,,,,source_only_missing_client_dimension,,2025-11-07,2026-02-19
client_770e8e5faa9cddfe,true,true,false,gsc_only,2025-06-24,2026-06-27,2025-11-17,
client_78d664d002f13e55,,,,source_only_missing_client_dimension,,
client_795153d5b7850ccf,true,true,false,gsc_only,2025-06-24,2026-06-27,2025-09-24,
client_7de9989c909e91a5,true,true,true,gsc_and_ga4,2026-04-07,2026-06-27,2026-05-18,2026-04-25
client_7eafe750768f0ac2,true,true,true,gsc_and_ga4,2026-02-23,2026-06-27,2026-02-26,2026-03-14
client_80ee5b7bd5f4eb89,true,true,false,gsc_only,2025-06-24,2026-06-27,,
client_810019792c9b8efc,true,true,true,gsc_and_ga4,2026-03-04,2026-06-27,2026-03-25,2026-03-26
client_835f9123c933bc01,,,,source_only_missing_client_dimension,,,2025-11-05,
client_861cdcccf8049915,false,false,false,no_search_or_analytics_access,2025-06-17,2026-06-29,2025-11-05,
client_86ebc2f12c01f586,true,true,true,gsc_and_ga4,2026-02-11,2026-06-27,2026-03-11,2026-03-10
client_895724a3631e810c,,,,source_only_missing_client_dimension,,
client_8ae2bfb5aa1ffa1e,true,true,false,gsc_only,2025-06-24,2026-06-27,2025-07-28,
client_8dbf3abdf07569e0,true,true,false,gsc_only,2025-05-26,2026-06-27,2025-07-29,
client_8ddc46da5414ffd8,true,true,false,gsc_only,2026-02-23,2026-06-27,2026-04-07,
client_91c8eb1698a7b352,true,true,false,gsc_only,2026-05-22,2026-06-27,,
client_95d66a740ef27635,false,false,false,no_search_or_analytics_access,2025-12-15,2026-06-29,,
client_95e0375ff4d52cf0,false,false,false,no_search_or_analytics_access,2026-04-16,2026-06-29,,
client_9958f0a7ae1df715,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-01-27,2025-10-29
client_9c26c096d6e57253,true,true,true,gsc_and_ga4,2026-03-23,2026-06-27,2026-04-29,2026-04-23
client_9d54435aabd95a6c,true,true,true,gsc_and_ga4,2025-12-12,2026-06-27,2026-02-19,2026-02-19
client_9f55cda45f6ca018,true,false,false,no_search_or_analytics_access,2026-06-09,2026-06-27,,
client_a1203ffecad62470,true,true,false,gsc_only,2025-06-24,2026-06-27,2025-10-20,
client_a22068e339bf95f5,true,true,true,gsc_and_ga4,2026-04-15,2026-06-27,2026-05-24,2026-05-22
client_a2eeb8899886adde,true,true,true,gsc_and_ga4,2025-06-24,2026-06-27,2025-07-06,2026-02-19
client_a60a11451483af1c,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-11-16,2026-02-19
client_a80fca3f171ed1de,false,true,true,gsc_and_ga4,2025-11-13,2026-06-29,2026-02-19,2026-03-09
client_ae27dc1eb540293f,true,false,false,no_search_or_analytics_access,2026-05-11,2026-06-27,,
client_aef6ffea193da149,true,true,true,gsc_and_ga4,2026-06-01,2026-06-27,2026-06-02,2026-05-29
client_b10cb2997d0c7c86,true,true,true,gsc_and_ga4,2025-06-16,2026-06-27,2025-06-18,2025-11-15
client_b2c982d4e2028c19,,,,source_only_missing_client_dimension,,
client_b2ce325550c5b03c,true,false,false,no_search_or_analytics_access,2026-06-01,2026-06-29,,
client_b77d0d5f08f05e64,true,true,true,gsc_and_ga4,2026-02-10,2026-06-27,2026-03-12,2026-03-19
client_ba5520003fdf9b10,,,,source_only_missing_client_dimension,,
client_ba65e80a1116ae41,false,true,true,gsc_and_ga4,2025-06-17,2026-06-29,2025-10-13,2025-11-09
client_bfc3117ed2a03e34,true,false,false,no_search_or_analytics_access,2026-04-15,2026-06-28,,
client_c182d11e4862a37d,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-06-21,2026-02-20
client_c353557474475e51,true,true,true,gsc_and_ga4,2026-03-27,2026-06-27,2026-04-30,2026-06-01
client_c3bfd4b85e1fef26,false,false,false,no_search_or_analytics_access,2025-05-26,2026-06-29,,
client_c7c2962f1c9c3089,true,true,true,gsc_and_ga4,2026-02-23,2026-06-27,2026-05-14,
client_ccdd78843409c8c7,true,true,true,gsc_and_ga4,2026-02-03,2026-06-27,2026-02-17,2026-02-20
client_cd12bcfd98942aa1,false,true,true,gsc_and_ga4,2025-10-15,2026-06-29,2025-10-20,2026-03-03
client_d03c3ff252a3c1fe,true,false,false,no_search_or_analytics_access,2026-04-09,2026-06-27,,
client_d211cb07b9059bab,true,true,true,gsc_and_ga4,2025-06-24,2026-06-27,2025-07-07,2026-02-19
client_d7df2f6a22182ed5,true,false,true,ga4_only,2025-05-26,2026-06-27,,
client_def0955f7a377868,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-07-17,2026-02-19
client_e00b29e582949543,true,true,true,gsc_and_ga4,2025-12-15,2026-06-27,2026-03-27,2026-03-25
client_e547b89c05043229,true,true,true,gsc_and_ga4,2025-05-26,2026-06-27,2025-11-15,2025-10-29
client_e5c2aa26a8598242,true,true,true,gsc_and_ga4,2025-12-12,2026-06-27,2026-02-19,2026-03-10
client_e6d6dce7c2fff733,true,true,true,gsc_and_ga4,2025-07-11,2026-06-27,,
client_e921cfa93fbe699b,,,,source_only_missing_client_dimension,,
client_f0ff30b229fe9b01,true,true,false,gsc_only,2025-05-26,2026-06-27,,
client_f4fa4e08a1d500d2,true,false,false,no_search_or_analytics_access,2026-06-25,2026-06-29,,
client_f623b01661d4bfe4,false,true,true,gsc_and_ga4,2025-06-20,2026-06-29,2026-01-09,2026-01-31
client_f63f09ff4e81aa58,true,false,false,no_search_or_analytics_access,2026-05-11,2026-06-27,,
"""

df = pd.read_csv(io.StringIO(data))
display(df.head())

,client_hash_id,is_active,has_gsc_access,has_ga4_access,access_profile,client_created_date,client_updated_date,gsc_data_start,ga4_data_start
0,client_04660893ae39614a,True,True,True,gsc_and_ga4,2026-04-15,2026-06-27,NaN,2026-05-22
1,client_05475c07ed21a83a,True,False,False,no_search_or_analytics_access,2026-04-01,2026-06-27,NaN,NaN
2,client_06d356715a8ff3b6,True,True,True,gsc_and_ga4,2026-03-23,2026-07-05,2026-04-10,2026-04-06
3,client_0797ff3a1fc9a6a5,True,False,False,no_search_or_analytics_access,2025-05-26,2026-06-27,2025-11-05,NaN
4,client_08a6a72ff48e62c0,True,True,False,gsc_only,2025-05-26,2026-06-27,2025-09-24,NaN


## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

The unit of analysis is a unique client, identified by `client_hash_id`. The time window for this data is determined by various date fields such as `client_created_date`, `client_updated_date`, `gsc_data_start`, and `ga4_data_start`.

In [8]:
# Verify the uniqueness of client_hash_id
num_unique_clients = df['client_hash_id'].nunique()
total_rows = len(df)

print(f"Number of unique clients: {num_unique_clients}")
print(f"Total number of rows: {total_rows}")

if num_unique_clients == total_rows:
    print("Each row represents a unique client.")
else:
    print("There are duplicate client_hash_ids.")

Number of unique clients: 100
Total number of rows: 100
Each row represents a unique client.


## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

### Field Categorization

Based on the column names and typical usage in a data contract, the fields are categorized as follows:

**Features (Independent Variables):**
*   `is_active`: A boolean indicating if the client is currently active.
*   `has_gsc_access`: A boolean indicating if the client has access to Google Search Console data.
*   `has_ga4_access`: A boolean indicating if the client has access to Google Analytics 4 data.
*   `access_profile`: A categorical string describing the type of access the client has (e.g., 'gsc_and_ga4', 'gsc_only', 'ga4_only', 'no_search_or_analytics_access').

**Labels (Dependent Variables):**
*   None explicitly defined in the provided dataset. A label would typically be a target variable for a predictive model.

**Context (Metadata/Descriptive Information):**
*   `client_hash_id`: A unique identifier for each client. Useful for tracking and joining, but not a feature for direct modeling.
*   `client_created_date`: The date when the client was created. Provides temporal context.
*   `client_updated_date`: The date when the client's information was last updated. Provides temporal context.
*   `gsc_data_start`: The start date of Google Search Console data availability for the client. Provides data coverage context.
*   `ga4_data_start`: The start date of Google Analytics 4 data availability for the client. Provides data coverage context.

**Excluded Fields:**
*   None are explicitly excluded at this stage, as all fields provide some level of information or context. Fields like `client_hash_id` might be dropped before modeling but are valuable for data integrity checks.

In [9]:
# Display data types to confirm initial assumptions
print("Data Types:")
print(df.dtypes)

print("\nValue Counts for Categorical Features:")
print("is_active:")
print(df['is_active'].value_counts(dropna=False))
print("\nhas_gsc_access:")
print(df['has_gsc_access'].value_counts(dropna=False))
print("\nhas_ga4_access:")
print(df['has_ga4_access'].value_counts(dropna=False))
print("\naccess_profile:")
print(df['access_profile'].value_counts(dropna=False))

Data Types:
client_hash_id         object
is_active              object
has_gsc_access         object
has_ga4_access         object
access_profile         object
client_created_date    object
client_updated_date    object
gsc_data_start         object
ga4_data_start         object
dtype: object

Value Counts for Categorical Features:
is_active:
is_active
True     73
False    18
NaN       9
Name: count, dtype: int64

has_gsc_access:
has_gsc_access
True     64
False    27
NaN       9
Name: count, dtype: int64

has_ga4_access:
has_ga4_access
True     51
False    40
NaN       9
Name: count, dtype: int64

access_profile:
access_profile
gsc_and_ga4                             50
no_search_or_analytics_access           26
gsc_only                                14
source_only_missing_client_dimension     9
ga4_only                                 1
Name: count, dtype: int64


### Data Type Conversion

To ensure proper analysis, we will convert the date columns to datetime objects and the boolean-like columns to actual boolean types. This will allow for correct date comparisons and filtering, and appropriate handling of boolean values.

In [10]:
# Convert date columns to datetime objects
date_cols = ['client_created_date', 'client_updated_date', 'gsc_data_start', 'ga4_data_start']
for col in date_cols:
    df[col] = pd.to_datetime(df[col], errors='coerce')

# Convert boolean columns to boolean type, handling 'NaN' values
bool_cols = ['is_active', 'has_gsc_access', 'has_ga4_access']
for col in bool_cols:
    # Map string 'true'/'false' to boolean, then convert to nullable boolean dtype
    df[col] = df[col].map({'true': True, 'false': False, True: True, False: False}).astype('boolean')

# Display data types again to confirm changes
print("Data Types after conversion:")
print(df.dtypes)

print("\nHead of DataFrame after conversion:")
display(df.head())

Data Types after conversion:
client_hash_id                 object
is_active                     boolean
has_gsc_access                boolean
has_ga4_access                boolean
access_profile                 object
client_created_date    datetime64[ns]
client_updated_date    datetime64[ns]
gsc_data_start         datetime64[ns]
ga4_data_start         datetime64[ns]
dtype: object

Head of DataFrame after conversion:


,client_hash_id,is_active,has_gsc_access,has_ga4_access,access_profile,client_created_date,client_updated_date,gsc_data_start,ga4_data_start
0,client_04660893ae39614a,True,True,True,gsc_and_ga4,2026-04-15,2026-06-27,NaT,2026-05-22
1,client_05475c07ed21a83a,True,False,False,no_search_or_analytics_access,2026-04-01,2026-06-27,NaT,NaT
2,client_06d356715a8ff3b6,True,True,True,gsc_and_ga4,2026-03-23,2026-07-05,2026-04-10,2026-04-06
3,client_0797ff3a1fc9a6a5,True,False,False,no_search_or_analytics_access,2025-05-26,2026-06-27,2025-11-05,NaT
4,client_08a6a72ff48e62c0,True,True,False,gsc_only,2025-05-26,2026-06-27,2025-09-24,NaT


### Missing Values

We will now inspect the number of missing values for each column to understand data completeness and identify any potential issues that may require imputation or special handling.

In [11]:
# Check for missing values
print("Missing values per column:")
print(df.isnull().sum())

print("\nPercentage of missing values per column:")
print((df.isnull().sum() / len(df)) * 100)

Missing values per column:
client_hash_id          0
is_active               9
has_gsc_access          9
has_ga4_access          9
access_profile          0
client_created_date     9
client_updated_date     8
gsc_data_start         36
ga4_data_start         53
dtype: int64

Percentage of missing values per column:
client_hash_id          0.0
is_active               9.0
has_gsc_access          9.0
has_ga4_access          9.0
access_profile          0.0
client_created_date     9.0
client_updated_date     8.0
gsc_data_start         36.0
ga4_data_start         53.0
dtype: float64


### Date Ranges

To confirm the time window of the data, we'll examine the minimum and maximum dates for each of the date columns.

In [12]:
# Check date ranges
date_cols = ['client_created_date', 'client_updated_date', 'gsc_data_start', 'ga4_data_start']
print("\nDate Ranges:")
for col in date_cols:
    print(f"{col}: Min = {df[col].min()}, Max = {df[col].max()}")


Date Ranges:
client_created_date: Min = 2025-05-26 00:00:00, Max = 2026-06-29 00:00:00
client_updated_date: Min = 2025-11-07 00:00:00, Max = 2026-07-05 00:00:00
gsc_data_start: Min = 2025-01-27 00:00:00, Max = 2026-06-02 00:00:00
ga4_data_start: Min = 2025-10-29 00:00:00, Max = 2026-06-01 00:00:00


## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

### Query Verification Summary

**Unit of Analysis:**
As verified by the code, `client_hash_id` is unique for each row, confirming that one row represents one unique client.

**Missing Values:**
- `is_active`, `has_gsc_access`, `has_ga4_access`, `client_created_date`: Each of these columns has 9 missing values (9% of the dataset).
- `client_updated_date`: Has 8 missing values (8% of the dataset).
- `gsc_data_start`: Has 36 missing values (36% of the dataset), indicating that GSC data start dates are often not available.
- `ga4_data_start`: Has 53 missing values (53% of the dataset), indicating that GA4 data start dates are frequently missing.

**Date Ranges (Overall Time Window):**
- `client_created_date`: Ranges from 2025-05-26 to 2026-06-29.
- `client_updated_date`: Ranges from 2025-11-07 to 2026-07-05.
- `gsc_data_start`: Ranges from 2025-01-27 to 2026-06-02.
- `ga4_data_start`: Ranges from 2025-10-29 to 2026-06-01.

These date ranges indicate that data collection spans from early 2025 to mid-2026, with variations in when specific client or data access records begin.

## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

### Data Limitations

This dataset, while informative, has several limitations that restrict the types of questions it can answer fully:

1.  **Incomplete Data History (Unbalanced History):** The high percentage of missing values in `gsc_data_start` (36%) and `ga4_data_start` (53%) means that for many clients, we do not know when their Search Console or Analytics 4 data became available. This prevents a complete understanding of data coverage over time for all clients.

2.  **Varied Data Availability for GSC/GA4:** The presence of `access_profile` values like 'gsc_only', 'ga4_only', and 'no_search_or_analytics_access', combined with the high missingness in `gsc_data_start` and `ga4_data_start`, indicates that not all clients have both GSC and GA4 data. This limits analyses requiring integrated data from both sources for all clients.

3.  **Missing Core Status Information:** The 9% missing values in `is_active`, `has_gsc_access`, `has_ga4_access`, and `client_created_date` means that for a subset of clients, we lack fundamental information about their status or when their records began. This can introduce bias if these missing values are not random.

4.  **Temporal Overlap Ambiguity:** While `client_created_date` and `client_updated_date` cover a broad period, the start dates for `gsc_data_start` and `ga4_data_start` vary significantly (e.g., GSC data starts as early as January 2025, while GA4 data starts no earlier than October 2025). This implies that concurrent data from both sources is not guaranteed for all clients or for the entire observed period, complicating longitudinal studies or comparisons between GSC and GA4 metrics.

5.  **"Source Only Missing Client Dimension" Clients:** The `access_profile` category 'source_only_missing_client_dimension' (9% of clients) suggests a class of clients for whom detailed access information might be incomplete or undefined in the data source itself. This group might need to be excluded from certain analyses or handled with specific imputation strategies.

6.  **No Explicit Labels:** As noted in the field categorization, there isn't an explicit label (target variable) in this dataset. This means the data cannot directly answer predictive questions like "which clients are likely to churn" or "which clients will increase their data access," without defining and deriving such labels from existing features.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.